Append region info to S3ToRedshitOperator if present

richardbadman opened the pull request:
It's possible to copy from S3 into Redshift across different regions, however, currently you are unable to do so with the S3ToRedshiftOperator. This PR simply makes this possible, by checking the aws connection passed has the region set in the extras part of the connection config. If this is set, it'll use this in line with the syntax defined [here](https://docs.aws.amazon.com/redshift/latest/dg/r_COPY.html)

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

richardbadman committed:
Append region info to S3ToRedshitOperator if present

It's possible to copy from S3 into Redshift across different regions,
however, currently you are unable to do so with the
S3ToRedshiftOperator. This PR simply makes this possible, by checking
the aws connection passed has the region set in the extras part of the
connection config. If this is set, it'll use this in line with the
syntax defined [here](https://docs.aws.amazon.com/redshift/latest/dg/r_COPY.html)

richardbadman committed:
Merge remote-tracking branch 'apache/main' into add-region-to-copy-statement

Adaverse commented on tests/providers/amazon/aws/transfers/test_s3_to_redshift.py:
```diff
@@ -324,6 +324,55 @@ def test_execute_role_arn(self, mock_run, mock_session, mock_connection, mock_ho
+
+        assert access_key in copy_query
+        assert secret_key in copy_query
+        assert extra["region"] in copy_query
```
Aren't the variables on the left and right in assertions hardcoded in the test? Meaning tests are bound to succeed. Aren't we supposed to test the return value of `_build_copy_query` with these assertions?

richardbadman replied:
Yeah, I found this odd when looking at the tests for this class _(but merely replicated the logic to be similar to all others)_, because all the tests have this. I can change it to do the assertions for the query that gets built from the actual test run, I think that would then satisfy the return value of `_build_copy_query`?

Adaverse replied:
Yes, i think the return value of `_build_copy_query` should cover the changes made with these assertions.

richardbadman replied:
Done, let me know if this is okay

richardbadman committed:
Update tests to make assertion checking valid

Following on from discussion in PR, currently the way assertion is done
is kind of redundant, as it's asserting static variables == other static
variables. Instead, this now gets compared to what gets generated from
the `_build_copy_query` function, this has been reflected for all
applicable test cases in this file.

Adaverse commented on tests/providers/amazon/aws/transfers/test_s3_to_redshift.py:
```diff
@@ -443,8 +492,21 @@ def test_using_redshift_data_api(self, mock_rs, mock_run, mock_session, mock_con
+
+        assert access_key in actual_copy_query
+        assert secret_key in actual_copy_query
         # test sql arg
```
this comment seems to be redundant since the next statement is clear on what it's doing.

Adaverse commented on tests/providers/amazon/aws/transfers/test_s3_to_redshift.py:
```diff
@@ -420,16 +477,8 @@ def test_using_redshift_data_api(self, mock_rs, mock_run, mock_session, mock_con
-        assert secret_key in copy_query
 
         mock_rs.execute_statement.assert_called_once()
         # test with all args besides sql
```
Same about this comment as well. IMO,  the comments are best suited for non-obvious or cheeky things, which is not the case above.

Adaverse reviewed (commented):
Things looks fine now.

richardbadman committed:
Remove redundant comments

Adaverse reviewed (approved):
except for static-checks, rest LGTM

richardbadman committed:
Follow static check formatting

potiuk commented on airflow/providers/amazon/aws/transfers/s3_to_redshift.py:
```diff
@@ -119,13 +119,16 @@ def __init__(
                     FROM 's3://{self.s3_bucket}/{self.s3_key}'
                     credentials
                     '{credentials_block}'
+                    {region_info}
```
Will that work when region is None? I srsly doubt

Adaverse replied:
Nice catch. It can only be None when user passes null in the extra config. Should be handled explicitly.

richardbadman replied:
Good spot, I've updated the parameter to not default to None which prevents the possibility of this happening

richardbadman committed:
Remove optional parameter & None reference from method
